### DISCUSSION TITLE
request.dynamic is reading first line of external bash script output only

### DISCUSSION THREAD

**Discussion by fedwa-zarrok** [savonet/liquidsoap#2269] [User Support]
Hi @toots, I am new to liquidsoap and icecast and I would like to create a dynamic playlist. I am using an external bash script to generate multiple paths to audio files, I am using request.dynamic for that but it seems like it only reads the first line of the output of my bash script. Can anyone please help me? Thanks

**_Here is my liquidsoap script :_**

```
set ("log.file.path","/home/admin/radio.log")

def my_request_function() = 

  result =
    list.hd(default="", get_process_lines("sh testScript.sh"))
  
  request.create(result, persistent=true)

end


m = request.dynamic(my_request_function)
m = audio_to_stereo(m)

radio = m

 
 clock.assign_new(id="/stream",[output.icecast(%vorbis(samplerate=44100, channels=2, quality=0.3),format="audio/ogg", fallible=true, host = "ip address", port = 8080 , password="password",  mount = "/test1",radio)])
```

**_Here is my bash script :_**

```
#!/bin/bash

now="$(date +'%Y-%m-%d')"

cd Playlist

cd Musique

cat $now*
```

**_And here is the output of bash script :_** 

```
./Audio/147/n.mp3 

./Audio/150/test.mp3
 
./Audio/308/eee.mp3
```

**Comment by rrolla** [comment]
Duplicate: https://github.com/savonet/liquidsoap/issues/2277

**Reply by toots** [reply]
Hi,

This is the intended behavior. `request.dynamic` processes requests one at a time. 

The best would be to adapt your callback logic to support pushing one request at a time.

You can optionally set the whole request queue at one in the callback but this is for advanced use and might prove tricky to do the right way.

Here's an example (haven't tested the code)
```ruby
# Function ref for callback function
# this is set after the source is created
set_queue_ref = ref(fun (_) -> ())
get_queue_ref = ref(fun () -> [])

def my_request_function() = 
  # Grab the requests
  requests = list.map(request.create, process.read.lines("sh testScript.sh"))

  # Grab the queue functions
  set = !set_queue_ref
  get = !get_queue_ref

  # Add new requests to the queue:
  queue = get()
  set([...queue, ...requests])

  # Return nothing here
  null()
end

# Define source
m = request.dynamic(my_request_function)

# Store actual callbacks:
set_queue_ref = m.set_queue
get_queue_ref = m.queue
```
